Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,461 @@
+presentation corresponding to a Coxeter matrix which is registered in a Coxeter system.
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries
+  equal to one and off-diagonal entries distinct from one.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup M`.
+* `equivCoxeterGroup` : Coxeter groups of isomorphic types are isomorphic.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter system.
+* `CoxeterMatrix.Aₙ`
+* `CoxeterMatrix.Bₙ`
+* `CoxeterMatrix.Dₙ`
+* `CoxeterMatrix.I₂ₘ`
+* `CoxeterMatrix.E₆`
+* `CoxeterMatrix.E₇`
+* `CoxeterMatrix.E₈`
+* `CoxeterMatrix.F₄`
+* `CoxeterMatrix.G₂`
+* `CoxeterMatrix.H₃`
+* `CoxeterMatrix.H₄`
+
+## References
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 4--5, 13--15
+
+## TODO
+
+* The canonical map from the type to the Coxeter group `W` is an injection.
+
+* A group `W` registered in a Coxeter system is a Coxeter group.
+
+* A Coxeter group is an instance of `IsCoxeterGroup`.
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+
+universe u
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries equal to one
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter : Prop where
+  symmetric : M.IsSymm := by aesop
+  diagonal : ∀ i : B, M i i = 1 := by aesop
+  off_diagonal : ∀ i j : B, i ≠ j → M i j ≠ 1 := by aesop
+
+namespace CoxeterGroup
+
+namespace Relations
+
+/-- The relations corresponding to a Coxeter matrix. -/
+def ofMatrix : B × B → FreeGroup B :=
+ Function.uncurry fun i j => (FreeGroup.of i * FreeGroup.of j) ^ M i j
+
+/-- The set of relations corresponding to a Coxeter matrix. -/
+def toSet : Set (FreeGroup B) :=
+  Set.range <| ofMatrix M
+
+end Relations
+
+end CoxeterGroup
+
+/-- The group presentation corresponding to a Coxeter matrix. -/
+def Matrix.CoxeterGroup := PresentedGroup <| CoxeterGroup.Relations.toSet M
+
+instance : Group (Matrix.CoxeterGroup M) := by
+  exact QuotientGroup.Quotient.group _
+
+namespace CoxeterGroup
+
+def of (b : B) : Matrix.CoxeterGroup M := PresentedGroup.of b
+
+@[simp]
+theorem of_apply (b : B) : of M b = PresentedGroup.of (rels := Relations.toSet M) b :=
+  rfl
+
+end CoxeterGroup
+
+/-- A Coxeter system `CoxeterSystem W` is a structure recording the isomorphism between
+a group `W` and the group presentation corresponding to a Coxeter matrix. Equivalently, this can
+be seen as a list of generators of `W` parameterized by the underlying type of `M`, which
+satisfy the relations of the Coxeter matrix `M`. -/
+structure CoxeterSystem (W : Type*) [Group W]  where
```

Comment:
The current order of variables seems to be `M W`. Is that intentional? I would have expected `W M`, I think.

Reply:
I was just allowing lean to set this since I am not passing in `M` explicitly here.  Also, since `M` is associated with the relations I would think it comes before the group that is generated by these relations (but maybe my thinking is backwards here?)

Out of curioiusity, why would have you expected `W M`?

Reply:
In analyogy with the `(W,S)` that you often see in the literature for Coxeter groups. But it's not very important, so feel free to keep `M W`.